### ISSUE TITLE
parseRM_GetNesting.pl - Nested elements are not merged in <out>.Nest.Corr.out

### ISSUE THREAD

**Issue by quantumdot** [4ureliek/Parsing-RepeatMasker-Outputs#4]
I would like to use your script `parseRM_GetNesting.pl` to generate a de-nested set of repetitive elements from repeatmasker input. It is critical for my purposes that any given element is reported only once in the output with corrected coordinates spanning all nested elements from a give insertion. This is because I would like to perform feature counting and intersection of sets.

I am struggling a little bit to understand the output of your scripts. Consider the output of the following commands using repeatmasker output from hg19 and mm9 genome builds, and the output generated by your script:
```bash
> grep LINE hg19_repeatmasker.4.0.5.txt | wc -l
1569213
> grep LINE hg19_repeatmasker.4.0.5.txt.Nest.Corr.out | wc -l
1791781
> grep LINE hg19_repeatmasker.4.0.5.txt.Nest.Only.out | wc -l
331916
> grep LINE hg19_repeatmasker.4.0.5.txt.Nest.OriginalLines.out | wc -l
646973
```
```bash
> grep LINE mm9_repeatmasker.txt | wc -l
977058
> grep LINE mm9_repeatmasker.txt.Nest.Corr.out | wc -l
1081806
> grep LINE mm9_repeatmasker.txt.Nest.Only.out | wc -l
163184
> grep LINE mm9_repeatmasker.txt.Nest.OriginalLines.out | wc -l
291800
```
The origional papers on the human and mouse genomes list the total number of LINE elements as [~868,000 for human](https://www.nature.com/nature/journal/v409/n6822/fig_tab/409860a0_T11.html) and [~660,000 for mouse](https://www.nature.com/nature/journal/v420/n6915/fig_tab/nature01262_T5.html). However the numbers reported by the output of your script do not match at all. In fact, it seems the numbers derived from `*.Corr.out` are even more greatly inflated than the original data. It appears that insertions disrupted by a nested insertion event are not merged into a single line as described in the documentation.

Consider the following lines of output from `mm9_repeatmasker.txt.Nest.Corr.out`:
```
Gname	Gstart	Gend	RealMaskedLen	Strand	Rname	Rclass/fam	Repstart	Rend	Repleft	%Div	%Del	%Ins	If-nest(ed/ing)	ID(blockNumber)

chr1	3000238	3000733	496	-	L1_Mur2	LINE/L1	(4488)	1389	913	21.4	11.4	4.5	na	na
chr1	3000238	3000792	522	-	L1_Mur2	LINE/L1	(4488)	1389	887	21.4946360153257	11.2107279693487	4.84367816091954	nesting	1
chr1	3000734	3000766	33	+	(TTTG)n	Simple_repeat	2	33	(0)	3.1	0.0	3.0	nested	1
chr1	3000767	3000792	26	-	L1_Mur2	LINE/L1	(6816)	912	887	23.3	7.6	11.4	na	na
chr1	3001288	3001583	296	-	Lx9	LINE/L1	(1596)	6048	5742	25.0	6.5	3.7	na	na
chr1	3001723	3002005	283	-	RLTR25A	LTR/ERVK	(0)	1028	625	19.7	33.2	0.7	na	na
chr1	3001723	3002615	847	-	RLTR25A	LTR/ERVK	(0)	1028	1	19.4336481700118	15.6873671782763	4.49551357733176	nesting	2
chr1	3002006	3002051	46	+	(TA)n	Simple_repeat	1	46	(0)	0.0	0.0	0.0	nested	2
chr1	3002052	3002615	564	-	RLTR25A	LTR/ERVK	(404)	624	1	19.3	6.9	6.4	na	na
chr1	3003213	3003261	49	-	L1M2	LINE/L1	(1343)	4800	4769	26.8	14.5	13.8	na	na
chr1	3003213	3003815	419	-	L1M2	LINE/L1	(1343)	4800	4378	26.8	14.5	13.8	nesting	3
chr1	3003262	3003445	184	-	B3	SINE/B2	(2)	214	4	20.1	14.7	0.0	nested	3
chr1	3003446	3003815	370	-	L1M2	LINE/L1	(1375)	4768	4378	26.8	14.5	13.8	na	na
```
Here there are three insertions that are disrupted by nesting: L1_Mur2 (lines 3, 4, 6); RLTR25A (lines 8, 9, 11); and L1M2 (lines 12, 13, 15), but according to the documentation should be merged into a single line. I would expect the output to look like the following:
```
name	Gstart	Gend	RealMaskedLen	Strand	Rname	Rclass/fam	Repstart	Rend	Repleft	%Div	%Del	%Ins	If-nest(ed/ing)	ID(blockNumber)

chr1	3000238	3000792	522	-	L1_Mur2	LINE/L1	(4488)	1389	887	21.4946360153257	11.2107279693487	4.84367816091954	nesting	1
chr1	3000734	3000766	33	+	(TTTG)n	Simple_repeat	2	33	(0)	3.1	0.0	3.0	nested	1
chr1	3001288	3001583	296	-	Lx9	LINE/L1	(1596)	6048	5742	25.0	6.5	3.7	na	na
chr1	3001723	3002615	847	-	RLTR25A	LTR/ERVK	(0)	1028	1	19.4336481700118	15.6873671782763	4.49551357733176	nesting	2
chr1	3002006	3002051	46	+	(TA)n	Simple_repeat	1	46	(0)	0.0	0.0	0.0	nested	2
chr1	3003213	3003815	419	-	L1M2	LINE/L1	(1343)	4800	4378	26.8	14.5	13.8	nesting	3
chr1	3003262	3003445	184	-	B3	SINE/B2	(2)	214	4	20.1	14.7	0.0	nested	3
```
Is the output in `*.Corr.out` correct?

**Comment by 4ureliek** [comment]
Hi,
Can you cc your input file corresponding to these lines? It looks like the original lines are printed in the 'corrected' output but I am not sure that it is the issue.
Also, the simple repeats should have been filtered out and not considered 'nested' elements...

**Comment by quantumdot** [comment]
My input file is the standard repeatmasker out file derived from the mm9 reference sequence obtainable [here](http://www.repeatmasker.org/genomes/mm9/RepeatMasker-rm328-db20090604/mm9.fa.out.gz).

The corresponding lines from my version of this file look like the following:
```
	SW	perc	perc	perc	query	position	in	query	matching	repeat	position	in	repeat
	score	div.	del.	ins.	sequence	begin	end	(left)	repeat	class/family	begin	end	(left)	ID

	687	17.4	0.0	0.0	chr1	3000002	3000156	(194195276)	C	L1_Mur2	LINE/L1	(4310)	1567	1413	1
	917	21.4	11.4	4.5	chr1	3000238	3000733	(194194699)	C	L1_Mur2	LINE/L1	(4488)	1389	913	1
	215	3.1	0.0	3.0	chr1	3000734	3000766	(194194666)	+	(TTTG)n	Simple_repeat	2	33	(0)	2
	845	23.3	7.6	11.4	chr1	3000767	3000792	(194194640)	C	L1_Mur2	LINE/L1	(6816)	912	887	1
	621	25.0	6.5	3.7	chr1	3001288	3001583	(194193849)	C	Lx9	LINE/L1	(1596)	6048	5742	3
	1320	19.7	33.2	0.7	chr1	3001723	3002005	(194193427)	C	RLTR25A	LTR/ERVK	(0)	1028	625	4
	414	0.0	0.0	0.0	chr1	3002006	3002051	(194193381)	+	(TA)n	Simple_repeat	1	46	(0)	5
	2352	19.3	6.9	6.4	chr1	3002052	3002615	(194192817)	C	RLTR25A	LTR/ERVK	(404)	624	1	4
	386	26.8	14.5	13.8	chr1	3003213	3003261	(194192171)	C	L1M2	LINE/L1	(1343)	4800	4769	6
	797	20.1	14.7	0.0	chr1	3003262	3003445	(194191987)	C	B3	SINE/B2	(2)	214	4	7
	386	26.8	14.5	13.8	chr1	3003446	3003815	(194191617)	C	L1M2	LINE/L1	(1375)	4768	4378	6
```

For my purposes, elements that are merged should only output the merged element rather than the merged element in addition to the original lines. Maybe you could make this an option if it was your intention to output the original lines in addition to the merged lines? 

Also, if I understand correctly, it should not be an issue for the simple repeats to be considered "nestED", but "nestING" could be an issue, however I don't believe that I observe the latter case.

**Comment by 4ureliek** [comment]
Hi,

So this code is 3y old and super messy - and it was not taking advantage of the block ID numbers generated by RepeatMasker (visually, it is the 'interrupted repeats' track on the UCSC genome browser).
I am writing a new script parseRM_merge_interrupted.pl - which in its current state outputs merged lines, like this:

Score | %Div | %Del | %Ins | Gname | Gstart | Gend | Gleft | Strand | Rname | Rclass/fam | Rstart(left) | Rend | Rend(left) | blockID | RealMaskedLen
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
10775.66667 | 11.6 | 1.3 | 2.4 | chr1 | 3000001 | 3004054 | nd | - | L1MdFanc_I | LINE/L1 | nd | nd | nd | 1 | 3009
3088 | 20.7 | 2.866666667 | 2.966666667 | chr1 | 3004041 | 3005441 | nd | + | L1_Rod | LINE/L1 | nd | nd | nd | 3 | 1337
5118 | 22.6 | 5.9 | 3.6 | chr1 | 3005461 | 3006764 | -192465207 | + | Lx9 | LINE/L1 | 6309 | 7644 | 0 | 5 |  
253 | 29 | 10.9 | 2.2 | chr1 | 3007005 | 3007230 | -192464741 | - | L1M3c | LINE/L1 | -4668 | 1473 | 1226 | 7 |  
2922.666667 | 17.33333333 | 6.633333333 | 7.866666667 | chr1 | 3008117 | 3010670 | nd | - | L1_Mur2 | LINE/L1 | nd | nd | nd | 8 | 1811
2917 | 20 | 7.7 | 1.2 | chr1 | 3008501 | 3009169 | -192462802 | + | RLTR26B_MM | LTR/ERVK | 1 | 716 | 0 | 9 |  

(slightly different since this is [mm10](http://www.repeatmasker.org/species/mm.html)), but similar lines as yours).
For now the -m option that would search again for nested elements is not implemented, but you should already get something much cleaner with the script I just pushed.

You are right, simple repeats cannot be nesting (they are 'added') - but unless you want to count them as well, it is a waste of space. It is easy to keep them if that's something people want, so I will make it an option.

**Comment by 4ureliek** [comment]
EDIT:
ended up uploading a script that will fit your needs - different than the one to get more details about the nesting/nested blocks structures.

**Comment by quantumdot** [comment]
Thank you very much @4ureliek!
